docs/proposals/18-shoot-CA-rotation.md

```diff
@@ -0,0 +1,195 @@
+- kubeconfig contains only `CA0`
+- `ServiceAccount` secrets contain only `CA0`
+- kubelet uses CLIENT CERT signed by `CA0`
+
+`t1`: User triggers first step of CA rotation process (--> phase one):
+
+- Generate `CA1`
+- Generate `CLIENT_CA1`
+- Create new `DNSRecord`, `Service`, Istio configuration, etc. for second `kube-apiserver` deployment
+- Deploy second `kube-apiserver` deployment trusting only CLIENT CERTS signed by `CLIENT_CA1` and using SERVER CERT signed by `CA1`
+- Update `kube-scheduler`, etc. to trust only CLIENT CERTS signed by `CLIENT_CA1` (`--client-ca-file` flag)
+- Update `kube-controller-manager` to issue new CLIENT CERTS with `CLIENT_CA1`
+- Update kubeconfig so that it points to the new `DNSRecord` and its CA bundle contains only `CA1` (if kubeconfig still contains a legacy CLIENT CERT then rotate the kubeconfig)
+- Update `ServiceAccount` secrets so that their CA bundle contains both `CA0` and `CA1`
+- Restart control plane components so that they point to the second `kube-apiserver` `Service` and so that their CA bundle contains only `CA1`
+- Renew CLIENT CERTS (sign them with `CLIENT_CA1`) for control plane components (Prometheus, DWD, legacy VPN) and point them to the second `kube-apiserver` `Service`
+- Adapt `apiserver-proxy-pod-mutator` to point `KUBERNETES_SERVICE_HOST` env variable to second `kube-apiserver`
+- Trigger node rollout
+  - This issues new CLIENT CERTS for all kubelets signed by `CLIENT_CA1` and points them to the second `DNSRecord`
+  - This restarts all `Pod`s and propagates `CA0` and `CA1` into their mounted `ServiceAccount` secrets
+- _Ask user to exchange all their client credentials (kubeconfig, CLIENT CERTS issued by `CertificateSigningRequest`s)_
+
+`t2`: User triggers second step of CA rotation process (--> phase two):
+
+- Update `ServiceAccount` secrets so that their CA bundle contains only `CA1`
+- Update `apiserver-proxy` to talk to second `kube-apiserver`
+- Drop first `DNSRecord`, `Service`, Istio configuration and first `kube-apiserver` deployment
+- Drop `CA0`
+- _Ask user to optionally restart their `Pod`s since they still contain `CA0` in memory._
+
+
+
+#### Advantages/Disadvantages approach two api servers
+
+- (+) User needs to adapt client credentials only once
+- (/) Unstable API server domain
+- (-) Probably more implementation effort
+- (-) More complex
+- (-) CA rotation process does not work similar for all CAs in our system
+
+#### Advantages/Disadvantages of currently preferred approach (see proposal)
+
+- (+) Implementation effort seems "straight-forward"
+- (+) CA rotation process works similar for all CAs in our system
+- (/) Stable API server domain
+- (-) User needs to adapt client credentials twice
+ 
+## Open Questions
+
+- Should we combine kubeconfig token rotation (not graceful today) with the 2-phase CA rotation process?
+- Do we need to mention the "`Pod`s still contain `CA0` in memory" part in the docs (GKE does not)?
```

Comment:
Google doesn't need to mention it, because it's not a problem they have, have they? They start the second batch of API servers on a new IP and roll with only the new CA in the belly (while the old workload that's not yet rolled has still the old CA talking to the old batch of API servers on the old IP).

Hmm... to be fair, we could mention it, so that people know (though only few will, as this is a difficult subject matter). Then again, if they are under attack, that's none of their concern. It's just the CA public key. If we are under attack and lost our CA private key, that's something else entirely. But then we (as well as Google) would need to actively inform all stakeholders and the difference is only, that Google can tell them to rotate once, while we need to tell them to rotate twice, but active communication is necessary either way. Is that about right?